presto-main/src/test/java/com/facebook/presto/sql/planner/TestPredicatePushdown.java

```diff
@@ -60,4 +63,142 @@ public void testPushDownToLhsOfSemiJoin()
+                                anyTree(
+                                        filter("LINE_ORDER_KEY > BIGINT '2'",
+                                                tableScan("lineitem", ImmutableMap.of(
+                                                        "LINE_ORDER_KEY", "orderkey",
+                                                        "LINE_QUANTITY", "quantity")))),
+                                anyTree(
+                                        filter("ORDERS_ORDER_KEY > BIGINT '2'",
+                                                tableScan("orders", ImmutableMap.of("ORDERS_ORDER_KEY", "orderkey")))))));
+    }
+
+    @Test
+    public void testEqualPredicateFromSourceSidePropagatesToFilterSideOfSemiJoin()
+    {
+        assertPlan("SELECT quantity FROM (SELECT * FROM lineitem WHERE orderkey IN (SELECT orderkey FROM orders) AND orderkey = 2)",
+                anyTree(
+                        semiJoin("LINE_ORDER_KEY", "ORDERS_ORDER_KEY", "SEMI_JOIN_RESULT",
+                                anyTree(
+                                        filter("LINE_ORDER_KEY = BIGINT '2' AND BIGINT '2' = LINE_ORDER_KEY", // TODO this should be simplified
+                                                tableScan("lineitem", ImmutableMap.of(
+                                                        "LINE_ORDER_KEY", "orderkey",
+                                                        "LINE_QUANTITY", "quantity")))),
+                                anyTree(
+                                        filter("ORDERS_ORDER_KEY = BIGINT '2' AND BIGINT '2' = ORDERS_ORDER_KEY", // TODO this should be simplified
+                                                tableScan("orders", ImmutableMap.of("ORDERS_ORDER_KEY", "orderkey")))))));
+    }
+
+    @Test
+    public void testPredicateFromSourceSideNotPropagatesToFilterSideOfSemiJoinIfNotIn()
+    {
+        assertPlan("SELECT quantity FROM (SELECT * FROM lineitem WHERE orderkey NOT IN (SELECT orderkey FROM orders) AND orderkey > 2)",
+                anyTree(
+                        semiJoin("LINE_ORDER_KEY", "ORDERS_ORDER_KEY", "SEMI_JOIN_RESULT",
+                                project(
+                                        filter("LINE_ORDER_KEY > BIGINT '2'",
+                                                tableScan("lineitem", ImmutableMap.of(
+                                                        "LINE_ORDER_KEY", "orderkey",
+                                                        "LINE_QUANTITY", "quantity")))),
+                                node(ExchangeNode.class, // NO filter here
+                                        project(
+                                                tableScan("orders", ImmutableMap.of("ORDERS_ORDER_KEY", "orderkey")))))));
+    }
+
+    @Test
+    public void testPredicateFromFilterSideNotPropagatesToSourceSideOfSemiJoinUsedInProjection()
+    {
+        assertPlan("SELECT orderkey IN (SELECT orderkey FROM orders WHERE orderkey > 2) FROM lineitem",
```

Comment:
add test with `...IN (SELECT orderkey FROM orders WHERE orderkey = 2)...`
and also maybe with:
```
`...IN (SELECT orderkey FROM orders WHERE orderkey = random(5)...`
```

Reply:
> add test with `...IN (SELECT orderkey FROM orders WHERE orderkey = 2)...`

There is test like this: `testEqualsPredicateFromFilterSidePropagatesToSourceSideOfSemiJoin`
It uses `SELECT quantity FROM (SELECT * FROM lineitem WHERE orderkey IN (SELECT orderkey FROM orders WHERE orderkey = 2))` query

> `...IN (SELECT orderkey FROM orders WHERE orderkey = random(5)...`

added